# Project 2

Jeff Lin   
300 599 323

As I am interested in the complexity and story telling of music, I naturally thought of comparing lyrics in modern day pop music vs broadway musicals, and want to see the difference in tone/depth of the different genera.   
I've also included 4 very different musicals as they very a lot in styles, they are: 
* Lion King
* Hamilton
* Book of Mormon
* Phantom of Opera

These can also be compared to each other, or indivudlly to pop lyrics

The lyrics are all collected from [Musical Lyrics](https://www.allmusicals.com/), whilst the top 10 ranking musicals was referenced from wiki [Broadway highest-grossing musicals](https://en.wikipedia.org/wiki/List_of_highest-grossing_musical_theatre_productions)

As we are just interested in the lyrics of the musical genera in general, so the accuracy of the list doesn't really matter long as it is well-known pieces of production, hence my selection

First I use the loader made previously to load the corpuses, setting up for the project

In [1]:
import sys
sys.path.append('../')
import corpus_loader


CORPUS_DIR = '../corpus/'
BILLBOARD = 'Billboard/'
BROADWAY = 'Broadway/'

Load all of the broadway corpuses

In [2]:
Lion_King_corpus, Lion_King_corpus_dict, Lion_King_corpus_df = corpus_loader.load_corpus(CORPUS_DIR+BROADWAY+'Lion King/')
Hamilton_corpus, Hamilton_corpus_dict, Hamilton_corpus_df = corpus_loader.load_corpus(CORPUS_DIR+BROADWAY+'Hamilton/')
Book_of_Mormon_corpus, Book_of_Mormon_corpus_dict, Book_of_Mormon_corpus_df = corpus_loader.load_corpus(CORPUS_DIR+BROADWAY+'Book of Mormon/')
Phantom_of_Opera_corpus, Phantom_of_Opera_corpus_dict, Phantom_of_Opera_corpus_df = corpus_loader.load_corpus(CORPUS_DIR+BROADWAY+'Phantom of Opera/')

Load the Billborad corpus for comparisons

In [3]:
Billboard_corpus, Billboard_corpus_dict, Billboard_corpus_df = corpus_loader.load_corpus(CORPUS_DIR+BILLBOARD+'2024/')

Check the corpus

In [4]:
print(Lion_King_corpus.fileids())
print(Lion_King_corpus.words())

['01 Circle of Life.txt', '02 Grasslands Chant.txt', '03 Morning Report.txt', '04 Lioness Hunt.txt', "05 I Just Can't Wait to Be King.txt", '06 Chow Down.txt', '07 They Live in You.txt', '08 Be Prepared.txt', '10 Rafiki Mourns.txt', '11 Hakuna Matata.txt', '12 One by One.txt', '13 Madness of King Scar.txt', '14 Shadowland.txt', '15 Lion Sleeps Tonight.txt', '16 Endless Night.txt', '17 Can You Feel the Love Tonight.txt', '18 He Lives in You (Reprise).txt', '20 King of Pride Rock-Circle of Life .txt']
['AFIKI', 'AND', 'ENSEMBLE', ':', 'Nants', 'ingonyama', ...]


In [5]:
print(Hamilton_corpus.fileids())
print(Hamilton_corpus.words())

['01 Alexander Hamilton.txt', '02 Aaron Burr.txt', '03 My Shot.txt', '04 The Story of Tonight.txt', '05 The Schuyler Sisters.txt', '06 Farmer Refuted.txt', "07 You'll Be Back.txt", '08 Right Hand Man.txt', "09 A Winter's Ball.txt", '10 Helpless.txt', '11 Satisfied.txt', '12 The Story of Tonight (Reprise).txt', '13 Wait For It.txt', '14 Stay Alive.txt', '15 Ten Duel Commandments.txt', '16 Meet Me Inside.txt', '17 That Would Be Enough.txt', '18 Guns and Ships.txt', '19 History Has Its Eye on You.txt', '20 Yorktown.txt', '21 What Comes Next.txt', '22 Dear Theodosia.txt', '23 Non-Stop.txt', "24 What'd I Miss.txt", '25 Cabinet Battle #1.txt', '26 Take a Break.txt', '27 Say No to This.txt', '28 The Room Where It Happens.txt', '29 Schuyler Defeated.txt', '30 Cabinet Battle #2.txt', '31 Washington on Your Side.txt', '32 One Last Time.txt', '33 I Know Him.txt', '34 The Adams Administration.txt', '35 We Know.txt', '36 Hurricane.txt', '37 The Reynolds Pamphlet.txt', '38 Burn.txt', '39 Blow Us All

In [6]:
print(Book_of_Mormon_corpus.fileids())
print(Book_of_Mormon_corpus.words())

['01 Hello.txt', '02 Two By Two.txt', '03 You And Me (But Mostly Me).txt', '04 Hasa Diga Eebowai.txt', '05 Turn It Off.txt', '06 I Am Here For You.txt', '07 All-American Prophet.txt', '08 Sal Tlay Ka Siti.txt', '09 Man Up.txt', '10 Making Things Up Again.txt', '11 Spooky Mormon Hell Dream.txt', '12 I Believe.txt', '13 Baptize Me.txt', '14 I Am Africa.txt', '15 Joseph Smith American Moses.txt', '16 Tomorrow Is A Latter Day.txt']
['ELDER', 'PRICE', ':', 'Hello', '!', 'My', 'name', ...]


In [7]:
print(Phantom_of_Opera_corpus.fileids())
print(Phantom_of_Opera_corpus.words())

['Phantom of Opera.txt']
['THE', 'STAGE', 'OF', 'THE', 'PARIS', 'OPERA', ',', ...]


In [8]:
print(Billboard_corpus.fileids())
print(Billboard_corpus.words())

['A Bar Song (Tipsy).txt', "Austin (Boots Stop Workin').txt", 'Beautiful Things.txt', 'Cruel Summer.txt', 'Espresso.txt', 'Lose Control.txt', 'Please Please Please.txt', 'Stick Season.txt', "TEXAS HOLD 'EM.txt", 'Too Sweet.txt']
['My', 'baby', 'want', 'a', 'Birkin', ',', 'she', "'", ...]


We can notice that whilst `Lion King` and `Book of Mormon` are imported as individule songs, `Hamilton` and `Phantom of Opera` are imported as the whole. This is partially to see the difference in methods, also it has a bit too many individule songs to import (around 80+)

Now, to first understand what we are working with, I have simplified the sub-genera of each of the four musicals:

| Musical | Sub-Genera | Theme |
|---------|-------------|----------|
| Lion King | Afro-disney; traditional-style show tunes | Uplifting; Circle of life; Disney |
| Hamilton  | hip hop; R&B; pop; soul; rap; traditional-style show tunes | Alexander Hamilton; American revolution |
| Book of Mormon | Classic Broadway; Traditional-style show tunes | Sarcastic/dark humer; Religion |
| Phantom of Opera | Classic Broadway; traditional-style show tunes; Operatic | Opera house; Gothic fiction |

Some of the information was consulted from the wikipedia pages of the musicals, some are just based on my music experices to categorise them. As the more important question we are interedted is the difference of the main genera `musical` vs `billboard`, I believe the `Hamilton` musical is a more appropriate comparison, as it shares the most similar amount of music sub-genera with each other, allowing for a better comparison with less noise or outside factor.

If I get to, I will still use the remaining 3 corpuses to compare and to see different results (if present).

### Initial asumptions and questions

* I am assuming that, as the nature of musical are to convey a full story / narative, vs the pop songs largely just conveying a single idea, it (musicals) will contain more informaion within its lyrics, thus leading to higher lexical diversity.

* Although the above hypothesis, I am unsure if it will nessecery lead to higher or perhaps, even less lexical diversity, as musicals also have very strong themes to them, it may lead to interesting scenarios where as cirtain names / ideas are repeated with much higher density than modern pop musics.

* There may be more questions along the way, but for now I am just going to start by using some simple metric analysis

First build a text analysisor 

In [9]:
# Create a custom word processor to process the text and return all variance of the processing outcome, 
# this is so I can compare the different processing methods and make sure all processing methods are 
# working correctly and as expected to all text in the corpus
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
from nltk.stem import WordNetLemmatizer
from nltk.tokenize import RegexpTokenizer
from nltk.probability import FreqDist

class WordProcessor:
    def __init__(self):
        self.lemmatizer = WordNetLemmatizer()
        self.stop_words = set(stopwords.words('english'))
        self.regex_tokenizer = RegexpTokenizer(r'\w+')
        self.tokenizer = word_tokenize
        
    def process_text(self, text) -> dict:
        """
        Process the text and return the tokenized, splitted, tokenized alphanumeric, 
        regex tokenized, regex tokenized without stop words, regex tokenized lematized, 
        regex tokenized without stop words and lematized

        Args:
            text (str): Text to be processed

        Returns:
            dict: Dictionary containing all the processed text
        """
        # processing methods
        tk = self.tokenizer(text)
        tk_ru_a = [word.lower() for word in tk if word.isalpha() and word.upper() != word]
        tk_ru_a_sw_lm = [self.lemmatizer.lemmatize(word) for word in tk_ru_a if word not in self.stop_words]

        text = text.lower()
        sp = text.split()
        tk = self.tokenizer(text)
        tk_a = [word for word in tk if word.isalpha()]
        rtk = self.regex_tokenizer.tokenize(text)
        tk_a_sw = [word for word in tk_a if word not in self.stop_words]
        rtk_sw = [word for word in rtk if word not in self.stop_words]
        tk_a_lm = [self.lemmatizer.lemmatize(word) for word in tk_a]
        rtk_lm = [self.lemmatizer.lemmatize(word) for word in rtk]
        tk_a_sw_lm = [self.lemmatizer.lemmatize(word) for word in tk_a_sw]
        rtk_sw_lm = [self.lemmatizer.lemmatize(word) for word in rtk_sw]


        # collect frequency distribution of rtk_lm and rtk_sw_lm
        fdist_rtk = FreqDist(rtk)
        fdist_rtk_lm = FreqDist(rtk_lm)
        fdist_rtk_sw_lm = FreqDist(rtk_sw_lm)
        fdist_tk_a_sw_lm = FreqDist(tk_a_sw_lm)
        fdist_tk_ru_a_sw_lm = FreqDist(tk_ru_a_sw_lm)

        # return the processed text
        return {
            'tk': tk,       # tokenized

            'sp': sp,       # splitted
            'tk_a': tk_a,   # tokenized alphanumeric
            'rtk': rtk,     # regex tokenized

            'tk_a_sw': tk_a_sw,       # tokenized alphanumeric without stop words
            'rtk_sw': rtk_sw,         # regex tokenized without stop words

            'tk_a_lm': tk_a_lm,       # tokenized alphanumeric lematized
            'rtk_lm': rtk_lm,         # regex tokenized lematized
            
            'tk_a_sw_lm': tk_a_sw_lm, # tokenized alphanumeric without stop words and lematized
            'rtk_sw_lm': rtk_sw_lm,   # regex tokenized without stop words and lematized

            'tk_ru_a': tk_ru_a,       # tokenized alphanumeric with full upper case ("character hints") removed 
            'tk_ru_a_sw_lm': tk_ru_a_sw_lm, # tokenized alphanumeric without stop words and lematized
        }, {
            'rtk': fdist_rtk,
            'rtk_lm': fdist_rtk_lm,
            'rtk_sw_lm': fdist_rtk_sw_lm,
            'tk_a_sw_lm': fdist_tk_a_sw_lm,
            'tk_ru_a_sw_lm': fdist_tk_ru_a_sw_lm,
        }

First check the metrics of the corpus, this is to get an idea of the text:

In [10]:
# get the token and type frequency distribution of the text
word_processor = WordProcessor()
hamilton_processed_text, hamilton_fdist_text = word_processor.process_text(' '.join(Hamilton_corpus.words()))
Billboard_corpus_text, Billboard_fdist_text = word_processor.process_text(' '.join(Billboard_corpus.words()))

raw = 'rtk'
process = 'tk_a_sw_lm'

print("Hamilton text")
print(hamilton_fdist_text[process])
print(hamilton_processed_text[process])
print(f"\nBillboard text")
print(Billboard_fdist_text[process])
print(Billboard_corpus_text[process])

Hamilton text
<FreqDist with 2564 samples and 18703 outcomes>
['burr', 'bastard', 'orphan', 'son', 'whore', 'scotsman', 'dropped', 'middle', 'forgotten', 'spot', 'caribbean', 'providence', 'impoverished', 'squalor', 'grow', 'hero', 'scholar', 'laurens', 'ten', 'dollar', 'founding', 'father', 'without', 'father', 'got', 'lot', 'farther', 'working', 'lot', 'harder', 'lot', 'smarter', 'self', 'starter', 'fourteen', 'placed', 'charge', 'trading', 'charter', 'jefferson', 'every', 'day', 'slave', 'slaughtered', 'carted', 'away', 'across', 'wave', 'struggled', 'kept', 'guard', 'inside', 'longing', 'something', 'part', 'brother', 'ready', 'beg', 'steal', 'borrow', 'barter', 'madison', 'hurricane', 'came', 'devastation', 'rained', 'man', 'saw', 'future', 'drip', 'dripping', 'drain', 'put', 'pencil', 'temple', 'connected', 'brain', 'wrote', 'first', 'refrain', 'testament', 'pain', 'burr', 'well', 'word', 'got', 'around', 'said', 'kid', 'insane', 'man', 'took', 'collection', 'send', 'mainland', '

Checking out the other 3 musicals just for reference

In [11]:
Lion_King_corpus_text, Lion_King_fdist_text = word_processor.process_text(' '.join(Lion_King_corpus.words()))
Book_of_Mormon_corpus_text, Book_of_Mormon_fdist_text = word_processor.process_text(' '.join(Book_of_Mormon_corpus.words()))
Phantom_of_Opera_corpus_text, Phantom_of_Opera_fdist_text = word_processor.process_text(' '.join(Phantom_of_Opera_corpus.words()))

raw = 'rtk'
process = 'tk_a_sw_lm'

print("Lion King text")
print(Lion_King_fdist_text[process])
print(Lion_King_corpus_text[process])
print(f"\nBook of Mormon text")
print(Book_of_Mormon_fdist_text[process])
print(Book_of_Mormon_corpus_text[process])
print(f"\nPhantom of Opera text")
print(Phantom_of_Opera_fdist_text[process])
print(Phantom_of_Opera_corpus_text[process])

Lion King text
<FreqDist with 803 samples and 2881 outcomes>
['afiki', 'ensemble', 'nants', 'ingonyama', 'bagithi', 'baba', 'sithi', 'uhm', 'ingonyama', 'nants', 'ingonyama', 'bagithi', 'baba', 'sithi', 'uhhmm', 'ingonyama', 'ingonyama', 'siyo', 'nqoba', 'ingonyama', 'ingonyama', 'nengw', 'enamabala', 'nants', 'ingonyama', 'bagithi', 'baba', 'sithi', 'uhm', 'ingonyama', 'nants', 'ingonyama', 'bagithi', 'baba', 'sithi', 'uhhmm', 'ingonyama', 'ingonyama', 'siyo', 'nqoba', 'ingonyama', 'ingonyama', 'nengw', 'enamabala', 'rafiki', 'day', 'arrive', 'planet', 'blinking', 'step', 'sun', 'see', 'ever', 'seen', 'ever', 'done', 'far', 'much', 'take', 'find', 'ever', 'found', 'sun', 'rolling', 'high', 'sapphire', 'sky', 'keep', 'great', 'small', 'endless', 'round', 'circle', 'life', 'move', 'u', 'despair', 'hope', 'faith', 'love', 'till', 'find', 'place', 'path', 'unwinding', 'circle', 'circle', 'life', 'rafiki', 'ensemble', 'circle', 'life', 'move', 'u', 'despair', 'hope', 'faith', 'love', 'till

Comparing the token counts and types of corpuses

In [12]:
import pandas as pd

types = pd.DataFrame(columns=['word raw type','word tokenize type', 'token count processed', 'laxical diversity'], index=['Billboard', 'Hamilton', 'Lion King', 'Book of Mormon', 'Phantom of Opera'])

raw = 'rtk'
process = 'tk_a_sw_lm'

types.loc['Billboard'] = [len(Billboard_fdist_text[raw]), len(Billboard_fdist_text[process]), len(Billboard_corpus_text[process]), len(Billboard_fdist_text[process])/len(Billboard_corpus_text[process])]
types.loc['Hamilton'] = [len(hamilton_fdist_text[raw]), len(hamilton_fdist_text[process]), len(hamilton_processed_text[process]), len(hamilton_fdist_text[process])/len(hamilton_processed_text[process])]
types.loc['Lion King'] = [len(Lion_King_fdist_text[raw]), len(Lion_King_fdist_text[process]), len(Lion_King_corpus_text[process]), len(Lion_King_fdist_text[process])/len(Lion_King_corpus_text[process])]
types.loc['Book of Mormon'] = [len(Book_of_Mormon_fdist_text[raw]), len(Book_of_Mormon_fdist_text[process]), len(Book_of_Mormon_corpus_text[process]), len(Book_of_Mormon_fdist_text[process])/len(Book_of_Mormon_corpus_text[process])]
types.loc['Phantom of Opera'] = [len(Phantom_of_Opera_fdist_text[raw]), len(Phantom_of_Opera_fdist_text[process]), len(Phantom_of_Opera_corpus_text[process]), len(Phantom_of_Opera_fdist_text[process])/len(Phantom_of_Opera_corpus_text[process])]

print(types.to_markdown()) 

|                  |   word raw type |   word tokenize type |   token count processed |   laxical diversity |
|:-----------------|----------------:|---------------------:|------------------------:|--------------------:|
| Billboard        |             677 |                  552 |                    1787 |            0.308898 |
| Hamilton         |            2980 |                 2564 |                   18703 |            0.13709  |
| Lion King        |             941 |                  803 |                    2881 |            0.278723 |
| Book of Mormon   |            1388 |                 1188 |                    4593 |            0.258654 |
| Phantom of Opera |            2092 |                 1797 |                    5552 |            0.323667 |


|                  |   word raw type |   word tokenize type |   token count processed |   laxical diversity |
|:-----------------|----------------:|---------------------:|------------------------:|--------------------:|
| Billboard        |             677 |                  552 |                    1787 |            0.308898 |
| Hamilton         |            2980 |                 2564 |                   18703 |            0.13709  |
| Lion King        |             941 |                  803 |                    2881 |            0.278723 |
| Book of Mormon   |            1388 |                 1188 |                    4593 |            0.258654 |
| Phantom of Opera |            2092 |                 1797 |                    5552 |            0.323667 |

Supricingly not suprices, the laxical diversity of musicals does not seem to have too much diversity from the billboards, apart from Hamilton, which seems dramatically less, this could perhaps be due to the way the texts are processed, so for now I'm going to look into more information, starting with frequency distribution of the corpus...

In [13]:
raw = 'rtk'
process = 'tk_a_sw_lm'

print("Billboard")
print(Billboard_fdist_text[raw].most_common(10))
print(Billboard_fdist_text[process].most_common(10))

print("Hamilton")
print(hamilton_fdist_text[raw].most_common(10))
print(hamilton_fdist_text[process].most_common(10))

Billboard
[('i', 224), ('you', 137), ('the', 121), ('it', 100), ('and', 98), ('s', 91), ('a', 82), ('me', 76), ('t', 74), ('that', 67)]
[('oh', 51), ('please', 37), ('know', 31), ('ooh', 24), ('come', 22), ('got', 20), ('baby', 19), ('take', 19), ('yeah', 19), ('right', 18)]
Hamilton
[('i', 1361), ('the', 1261), ('you', 1025), ('to', 787), ('hamilton', 739), ('a', 732), ('and', 707), ('it', 506), ('s', 491), ('burr', 475)]
[('hamilton', 739), ('burr', 475), ('washington', 234), ('company', 223), ('eliza', 205), ('da', 199), ('lafayette', 178), ('laurens', 166), ('jefferson', 158), ('angelica', 152)]


We see the top words are all names, but maybe with a bit too much frequency then they should have, going to check the source text to see what it looks like

In [17]:
# print the first 50 words of Hamilton
print(Hamilton_corpus_dict['01 Alexander Hamilton'][:1000])

[BURR]
How does a bastard, orphan, son of a whore and a Scotsman, dropped in the middle of a forgotten
spot in the Caribbean by Providence impoverished in squalor
grow up to be a hero and a scholar?

[LAURENS]
The Ten dollar, founding father without a father
got a lot farther by working a lot harder
by being a lot smarter by being a self-starter
by fourteen, they placed him in charge of a trading charter

[JEFFERSON]
And every day while slaves were being slaughtered and carted away
across the waves he struggled and kept his guard up
Inside he was longing for something to be a part of
the brother was ready to beg steal borrow or barter

[MADISON]
Then a hurricane came and devastation rained
our man saw his future drip-dripping down the drain
put a pencil to his temple connected it to his brain
and he wrote his first refrain a testament to his pain

[BURR]
Well the word got around they said this kid is insane man
took up a collection just to send him to the mainland
Get your education do

from this we can see, the top frequent words are actually just name titles that indicates who's line it is to sing, and by looking into the source text we can see that it is expressed by all caps, which I think I may be able to remove via conditions during the word processing process.

In [21]:
process1 = 'tk_a_sw_lm'
process2 = 'tk_ru_a_sw_lm'


print("Hamilton")
print(hamilton_fdist_text[process1].most_common(20))
print(hamilton_fdist_text[process2].most_common(20))

# print lexical diversity 
print(f'\nLaxical diversity')
print(f"pre-process:  {len(hamilton_fdist_text[process1])/len(hamilton_processed_text[process1]):.4f}")
print(f"post-process: {len(hamilton_fdist_text[process2])/len(hamilton_processed_text[process2]):.4f}")

Hamilton
[('hamilton', 739), ('burr', 475), ('washington', 234), ('company', 223), ('eliza', 205), ('da', 199), ('lafayette', 178), ('laurens', 166), ('jefferson', 158), ('angelica', 152), ('know', 148), ('ensemble', 147), ('wait', 132), ('mulligan', 127), ('take', 122), ('time', 119), ('away', 115), ('like', 115), ('look', 114), ('get', 112)]
[('da', 199), ('know', 148), ('wait', 132), ('hamilton', 128), ('take', 122), ('time', 119), ('away', 115), ('like', 115), ('look', 114), ('get', 112), ('shot', 111), ('hey', 111), ('say', 101), ('got', 98), ('see', 95), ('go', 95), ('never', 94), ('burr', 93), ('let', 93), ('one', 93)]

Laxical diversity
pre-process:  0.1371
post-process: 0.1604


from the above processing, we can see that it succesfully removes the noise from character hints, and more accurately reflects the proper context of the corpus, however it also appears the word `da` is also like filler words, but as the billboard corpus also has problems with words like `oh / woo / yeah / whoa / ya`, and debating with the nature of lyrics I decided to leave them in, as it is also a delibrate musical/lyrical choice to have these words as phrases, as they serve a "function" in which it envoke emotions